inference-explainers/inference_api.md

```diff
@@ -0,0 +1,252 @@
+**Authors:** <br>
+[Steve Gan][5], Google Privacy Sandbox <br>
+
+## Introduction
+
+This document provides a comprehensive overview of the Bidding and Auction (B&A) Inference service API with externally supplied models ([B&A inference overview][1]), including supported machine learning (ML) frameworks, model formats, model management processes, and the JavaScript API for executing inference tasks.
+
+## Supported Machine Learning Frameworks and Model Formats
+
+The Tensorflow and PyTorch runtimes are supported, each with a dedicated inference service sidecar binary. Currently, we offer the Tensorflow v2.14.0 sidecar binary and the PyTorch v2.1.1 sidecar binary.
+
+For Tensorflow, the SavedModel format is used while PyTorch uses the TorchScript format. We recommend that ad techs develop and train models in an environment in which the ML runtime version aligns with the intended sidecar.
+
+The maximum size for a single ML model is 2GB.
+
+## Model management
+
+ML models used by the inference service are fetched periodically from a linked cloud storage (e.g., [Google Cloud Storage buckets][2], [Amazon S3 buckets][3]) and made available for serving. Each model is uniquely identified by its path within the cloud storage. Ad techs can implement model versioning by structuring model storage paths to include version identifiers. For example, storing models at paths such as “pcvr_v1” and “pcvr_v2” distinguishes between the two versions.
+
+Ad techs can prepare a model configuration file in JSON format within the same cloud storage as the models. The B&A service periodically checks the configuration file (according to period configurable by ad techs) for any changes and triggers the loading of new models into the inference service sidecar’s memory as needed.
+
+For example:
+
+```json
+{
+  "model_metadata": [
+    {
+      "model_path": "pcvr_v1",
+      "checksum": "dd94b3b08ea19b4240aa5e5f68ff0447b40e37ebdd06cc76fa1a2cf61143a10c",
+      "warm_up_batch_request_json": "..."
+    },
+    {
+      "model_path": "pcvr_v2",
+      "checksum": "...",
+      "warm_up_batch_request_json": "..."
+    }
+  ]
+}
+```
+
+This model configuration file features a top-level array with each entry containing the metadata of the fetched models. The mandatory **model_path** field specifies the path to the model in the cloud storage. The optional **checksum** field is the SHA256 checksum of the model represented as a hexadecimal string.
+
+Model checksums are computed using the following steps:
+
+1. Compute the SHA256 checksum for each individual model file.
+2. Arrange the file checksums in ascending order based on their file paths.
+3. Concatenate all the ordered file checksums into a single string.
+4. Compute the final SHA256 checksum on the concatenated checksum string.
+
+The listed steps are equivalent to the following bash command:
+
+```
+find <model_path> -type f -exec sha256sum {} \; | sort -k 2 | awk '{print $1}' | tr -d '\n' | sha256sum | awk '{print $1}'
```

Comment:
We actually don't need to. This command already works for both cases. Because the find command checks the directory entries, it can figure out if the model path is a single file or a directory and performs a match only if a corresponding entry is found. We ran into the issue in our case because CPIO lacks the notion of directory.

More concretely, if model is a directory, model1 is a file or directory, "find model -type f" only finds model/file1 and model/file2 because in the current directory, there is an entry for "model" which is a directory. It then recursively traverses "model". "find model/ -type f" produces the same result. 

If model is a regular file. "find model -type f" finds the regular file only. In both cases, model1 is ignored.

Reply:
Done